src/scripts/mod.rs

```diff
@@ -0,0 +1,162 @@
+}
+
+impl Script {
+    pub fn build(
+        command: Executable,
+        path: Option<PathBuf>,
+        arguments: Vec<String>,
+        ip: IpAddr,
+        ports: Vec<u16>,
+        tags: Option<Vec<String>>,
+    ) -> Self {
+        let mut command_match = String::from("nmap");
+        match command {
+            Nmap => command_match = "nmap".into(),
+            _ => {}
+        }
+
+        // Main command to use ex. nmap, python, perl, bash
+        Self {
+            // command like nmap, or python
+            command: command_match,
+            // command: command,
+
+            // optional path to the executable
+            path: path,
+
+            // user given arguments on commandline after --
+            arguments: arguments,
+            ip: ip,
+            ports: ports,
+
+            /*  script tags stored in file header? We can use toml to read it into a struct after parsing the headers manually first
+                so read file, parse first 5 lines, remove first line (shebang) then throw it to toml to parse the tags to the struct.
+
+
+            Shebang is always the first line in a script file (if it's there at all), and we can also check for it to check the interpreter
+            example:
+
+            #!/bin/python or #!/bin/perl or #!/bin/sh or #!/bin/python3
+            # tags = [core_approved, http, robotstxt, hyper, get_request]
+            # ports = [80]
+            # name = "Robots.txt grabbing"
+            # description = "Grabs the robots.txt file from webservers"
+
+            fn main()     // start of the script in the file
+            */
+            tags: tags,
+        }
+    }
+
+    pub fn run(mut self) -> Result<String> {
+        match &self.command {
+            Nmap => {
+                let mut arguments = vec!["nmap".to_string()];
+                arguments = self.arguments.iter().map(|arg| arg.into()).collect();
+                let nmap_str_ports: Vec<String> =
+                    self.ports.iter().map(|port| port.to_string()).collect();
+                let ports_str = nmap_str_ports.join(",");
+
+                arguments.push("-vvv".into());
+                if self.ip.is_ipv6() {
+                    arguments.push("-6".into());
+                }
+                arguments.push("-p".into());
+                arguments.push(ports_str);
+                arguments.push(self.ip.to_string());
+                self.arguments = arguments;
+
+                println!("nmap {:?}", self.arguments);
+                match self.execute_script() {
+                    Ok(result) => Ok(result),
+                    Err(e) => Err(e),
+                }
+            }
+            Python => {
+                todo!();
+            }
+        }
+    }
+
+    fn execute_script(self) -> Result<String> {
+        let process = Exec::cmd(self.command).args(&self.arguments);
+        match process.capture() {
+            Ok(c) => {
+                let es = match c.exit_status {
```

Comment:
Python scripts, by default, do not exit with OS status codes by the way -- we'd have to enforce this with the users submissions :)

Reply:
I think as long as it completes without error, it will return a `0`, every program ends with an exit systemcall, even if it is the interpreter. We should make some tests to see how it reacts to success and errors and what gets returned on the end. I will do that.